src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,296 @@
+    }
+
+    // Check output_index
+    int output_index = params[2].get_int();
+    if (output_index < 0 || output_index >= ZC_NUM_JS_OUTPUTS) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid output_index");
+    }
+
+    // Get message if it exists
+    string msg;
+    if (params.size() == 4) {
+        msg = params[3].get_str();
+    }
+
+    // Create PaymentDisclosureKey
+    PaymentDisclosureKey key = {hash, (size_t)js_index, (uint8_t)output_index };
+
+    // TODO: In future, perhaps init the DB in init.cpp
+    shared_ptr<PaymentDisclosureDB> db = PaymentDisclosureDB::sharedInstance();
+    PaymentDisclosureInfo info;
+    if (!db->Get(key, info)) {
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Could not find payment disclosure info for the given joinsplit output");
+    }
+
+    PaymentDisclosure pd( wtx.joinSplitPubKey, key, info, msg );
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << pd;
+    string strHex = HexStr(ss.begin(), ss.end());
+    return strHex;
+}
+
+
+
+/**
+ * RPC call to validate a payment disclosure data blob.
+ */
+UniValue z_validatepaymentdisclosure(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    auto fEnablePaymentDisclosure = fExperimentalMode && GetBoolArg("-paymentdisclosure", false);
+    string strPaymentDisclosureDisabledMsg = "";
+    if (!fEnablePaymentDisclosure) {
+        strPaymentDisclosureDisabledMsg = "\nWARNING: Payment disclosure is curretly DISABLED. This call always fails.\n";
+    }
+
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "z_validatepaymentdisclosure \"paymentdisclosure\"\n"
+            "\nValidates a payment disclosure.\n"
+            "\nEXPERIMENTAL FEATURE\n"
+            + strPaymentDisclosureDisabledMsg +
+            "\nArguments:\n"
+            "1. \"paymentdisclosure\"     (string, required) Hex data string\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_validatepaymentdisclosure", "\"hexblob\"")
+            + HelpExampleRpc("z_validatepaymentdisclosure", "\"hexblob\"")
+        );
+
+    if (!fEnablePaymentDisclosure) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: payment disclosure is disabled.");
+    }
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+
+    string hexInput = params[0].get_str();
+    if (!IsHex(hexInput))
+    {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected payment disclosure data in hexadecimal format.");
+    }
+
+    // Unserialize the payment disclosure data into an object
+    PaymentDisclosure pd;
+    CDataStream ss(ParseHex(hexInput), SER_NETWORK, PROTOCOL_VERSION);
+    try {
+        ss >> pd;
+        // too much data is ignored, but if not enough data, exception of type ios_base::failure is thrown,
+        // CBaseDataStream::read(): end of data: iostream error
```

Comment:
Make sure that an unrecognized version (or marker, if that is added) is reported.

Reply:
Oh, it already is for the version.